### PR TITLE
fix: locale-independent float parsing (fixes docling#1455)

### PR DESCRIPTION
**Pull request by evb87-tech**
## Summary

Replace all locale-sensitive floating-point parsing (`std::stod`, `atof` via QPDF) with locale-independent `std::from_chars` (C++17/20). This fixes PDF coordinate corruption under locales that use `,` as the decimal separator (French, German, Portuguese, ~40 others).

- New `utils/numeric.h` header with `locale_safe_stod()` and `locale_safe_numeric_value()` — drop-in replacements that use `std::from_chars`
- All 9 `std::stod()` call sites replaced in `stream_decoder.h` and `base_font.h`
- All 7 `getNumericValue()` call sites wrapped in `stream_instruction.h`, `page_dimension.h`, `page.h`, `grph.h`, `text.h`, `to_json.h`
- Python `__init__.py` sets `LC_NUMERIC=C` at import time as defence-in-depth
- Comprehensive test suite (`test_locale_safety.py`) with 12 tests across unit/integration/regression

Fixes: https://github.com/docling-project/docling/issues/1455

## Problem

Under locales where `LC_NUMERIC` uses `,` as the decimal separator (e.g. `fr_FR.UTF-8`), **every** floating-point value extracted from a PDF is silently corrupted:

```
std::stod("72.5")   → 72.0   (period not recognised, parsing stops)
atof("595.276")     → 595.0  (same issue, inside QPDF)
```

This affects page dimensions, text coordinates, font metrics, graphics state — the entire parsing pipeline. The result is garbled layout or hangs in downstream processing.

### Root cause (two independent paths)

| Layer | Function | Location | Issue |
|-------|----------|----------|-------|
| QPDF (upstream) | `atof()` in `getNumericValue()` | `libqpdf/QPDFObjectHandle.cc` | Called from 7 sites in docling-parse |
| docling-parse | `std::stod()` | `stream_decoder.h`, `base_font.h` | 9 direct calls |

Both `atof()` and `std::stod()` respect `LC_NUMERIC` per the C/C++ standard.

### 1. `utils/numeric.h` (new file)

Two locale-safe utilities:

```cpp
// Drop-in replacement for std::stod()
inline double locale_safe_stod(const std::string& str) {
    double value = 0.0;
    auto [ptr, ec] = std::from_chars(str.data(), str.data() + str.size(), value);
    // ... error handling matching std::stod contract
    return value;
}

// Safe wrapper for QPDF's getNumericValue()
inline double locale_safe_numeric_value(QPDFObjectHandle& obj) {
    if (obj.isInteger())
        return static_cast<double>(obj.getIntValue());  // integers are safe
    // For reals: re-parse from string representation, bypassing QPDF's atof()
    return locale_safe_stod(obj.getRealValue());
}
```

`std::from_chars` is guaranteed locale-independent by C++17 (double support in C++20, available in GCC 11+, Clang 16+, MSVC 19.24+). The project already uses C++20.

### 2. All call sites replaced

| File | Calls replaced | Type |
|------|---------------|------|
| `stream_decoder.h` | 3 | `std::stod` → `locale_safe_stod` |
| `base_font.h` | 6 | `std::stod` → `locale_safe_stod` |
| `stream_instruction.h` | 1 | `getNumericValue` → `locale_safe_numeric_value` |
| `page_dimension.h` | 1 | `getNumericValue` → `locale_safe_numeric_value` |
| `page.h` | 3 | `getNumericValue` → `locale_safe_numeric_value` |
| `grph.h` | 1 | `getNumericValue` → `locale_safe_numeric_value` |
| `text.h` | 1 | `getNumericValue` → `locale_safe_numeric_value` |
| `to_json.h` | 1 | `getNumericValue` → `locale_safe_numeric_value` |

### 3. Python defence-in-depth

`docling_parse/__init__.py` now forces `LC_NUMERIC=C` at import time — same approach used by Poppler and Cairo. This catches any codepath we may have missed, including QPDF internals.

## Test plan

New test file: `tests/test_locale_safety.py` — 12 tests in 5 classes:

### Unit tests (`TestLocaleIndependentParsing`)
- [x] Page dimensions not truncated under comma locale
- [x] Fractional dimensions retained (e.g. 595.276 not → 595)
- [x] Text cell coordinates identical across locales

### Integration tests (`TestFullPipelineLocaleResilience`)
- [x] All 39 regression PDFs parse without error under hostile locale
- [x] Page count consistent across locale switches
- [x] Shape coordinates (graphics state) locale-independent

### Edge-case tests (`TestNumericEdgeCases`)
- [x] BytesIO loading works under comma locale
- [x] All pages of multi-page PDFs valid under hostile locale
- [x] Widget and hyperlink bounding boxes correct under comma locale

### Regression tests (`TestCoordinateReproducibility`)
- [x] Coordinates bit-identical across repeated locale switches (4 iterations, alternating C/fr_FR)
- [x] Text content extraction unaffected by locale

### Python guard tests (`TestPythonLocaleGuard`)
- [x] Import resets LC_NUMERIC when set to comma locale
- [x] Guard is no-op under C locale

Tests exercise `fr_FR`, `de_DE`, `pt_BR`, `es_ES`, `it_IT` — using whichever is available on the system.

## Compatibility

- **C++ standard**: Uses `std::from_chars` (C++17/20) — project already targets C++20
- **No API changes**: all changes are internal implementation details
- **No new dependencies**: `<charconv>` is part of the C++ standard library
- **Backwards compatible**: behaviour is identical for the default C/POSIX locale

## Related

- Upstream QPDF bug: `atof()` in `QPDFObjectHandle::getNumericValue()` should use `std::from_chars` or their existing `QUtil` locale-safe utilities. Worth filing separately on `qpdf/qpdf`.

### CODE REVIEW

**Review by PeterStaar-IBM — CHANGES_REQUESTED** [review]

File: src/parse/utils/numeric.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    //
+    // This function is a drop-in replacement for obj.getNumericValue()
+    // anywhere a QPDFObjectHandle is known to be a number.
+    inline double locale_safe_numeric_value(QPDFObjectHandle& obj)
+    {
+      if (obj.isInteger())
+        {
+          return static_cast<double>(obj.getIntValue());
+        }
+
```
[the remark this answers is not in the document]
Comment (evb87-tech) [reply src/parse/utils/numeric.h]: Done in af2fe50 — added the explicit `else if (obj.isReal())` branch and the `else` throw.

One small substitution: I used `std::invalid_argument` instead of `std::value_error`, since `value_error` isn't in the C++ standard library (Python's `ValueError` is the closest analogue in spirit). It's also the exception type already thrown by `locale_safe_stod` for parse failures in this same file, so the exception types stay consistent.

Merge conflict with main is resolved. Tests pass locally (43/43, including the 13 new locale-safety tests).

**Comment by PeterStaar-IBM** [comment]
@evb87-tech Seems like we have some compilation errors, can you have a quick check?

**Comment by evb87-tech** [comment]
Thanks for the heads up. I checked the failing wheel jobs: the compilation errors came from `src/parse/utils/numeric.h` using `std::from_chars(first, last, double&)`. That overload is not available across the full wheel matrix: macOS arm64 libc++ reports it as deleted, and the py3.10 manylinux image uses GCC/libstdc++ 10 where `from_chars` is integer-only.

I pushed `7587f21` to replace the floating `from_chars` call with a `std::istringstream` imbued with `std::locale::classic()`, which keeps the parsing independent of `LC_NUMERIC` without requiring newer standard-library support.

Local verification passed:
- `make -C build -j 4`
- `make -C build install -j 4`
- `uv run pytest tests/test_locale_safety.py` (`13 passed`)

CI should now rerun on the pushed commit.

**Comment by PeterStaar-IBM** [comment]
> Thanks for the heads up. I checked the failing wheel jobs: the compilation errors came from `src/parse/utils/numeric.h` using `std::from_chars(first, last, double&)`. That overload is not available across the full wheel matrix: macOS arm64 libc++ reports it as deleted, and the py3.10 manylinux image uses GCC/libstdc++ 10 where `from_chars` is integer-only.
> 
> I pushed `7587f21` to replace the floating `from_chars` call with a `std::istringstream` imbued with `std::locale::classic()`, which keeps the parsing independent of `LC_NUMERIC` without requiring newer standard-library support.
> 
> Local verification passed:
> 
> * `make -C build -j 4`
> * `make -C build install -j 4`
> * `uv run pytest tests/test_locale_safety.py` (`13 passed`)
> 
> CI should now rerun on the pushed commit.

excellent, let's run the CI again!

**Comment by evb87-tech** [comment]
seems we're good now ! @PeterStaar-IBM

**Review by PeterStaar-IBM — APPROVED** [review]
lgtm!
